pkg/build/generator/generator.go

```diff
@@ -267,60 +267,64 @@ func (g *BuildGenerator) generateBuildFromConfig(ctx kapi.Context, bc *buildapi.
 		}
 		updateCustomImageEnv(build.Parameters.Strategy.CustomStrategy, image)
 	}
+
+	// lookup the destination from the referenced image repository
+	if build.Parameters.Output.To != nil {
+		if build.Parameters.Output.To.Kind == "ImageStreamImage" {
+			return nil, fatalError{fmt.Errorf("ImageStreamImage is not allowed in Output.To")}
+		}
+		image, err := g.resolveImageStreamReference(ctx, build.Parameters.Output.To, bc.Namespace)
+		if err != nil {
+			return nil, err
+		}
+		build.Parameters.Output.To = &kapi.ObjectReference{
+			Kind: "DockerImage",
+			Name: image,
+		}
+	}
+
 	return build, nil
 }
 
 // resolveImageStreamReference looks up the ImageStream[Tag/Image] and converts it to a
 // docker pull spec that can be used in an Image field.
-func (g *BuildGenerator) resolveImageStreamReference(ctx kapi.Context, from *kapi.ObjectReference, defaultNamespace string) (string, error) {
+func (g *BuildGenerator) resolveImageStreamReference(ctx kapi.Context, ref *kapi.ObjectReference, defaultNamespace string) (string, error) {
 	var namespace string
-	if len(from.Namespace) != 0 {
-		namespace = from.Namespace
+	if len(ref.Namespace) != 0 {
+		namespace = ref.Namespace
 	} else {
 		namespace = defaultNamespace
 	}
 
-	glog.V(4).Infof("Resolving ImageStreamReference %s of Kind %s in namespace %s", from.Name, from.Kind, namespace)
-	switch from.Kind {
-	case "ImageStream":
-		// NOTE: The 'ImageStream' reference should be used only for the 'output' image
-		is, err := g.Client.GetImageStream(kapi.WithNamespace(ctx, namespace), from.Name)
-		if err != nil {
-			glog.V(2).Infof("Error getting ImageStream %s/%s: %v", namespace, from.Name, err)
-			return "", err
-		}
-		image, err := imageapi.DockerImageReferenceForStream(is)
-		if err != nil {
-			glog.V(2).Infof("Error resolving Docker image reference for %s/%s: %v", namespace, from.Name, err)
-			return "", err
-		}
-		return image.String(), nil
+	glog.V(4).Infof("Resolving ImageStreamReference %s of Kind %s in namespace %s", ref.Name, ref.Kind, namespace)
```

Comment:
looking at the conversions  and talking to @smarterclayton we are still using ImageStream as the internal Kind for a To field.  So i don't think you can remove that case from this switch.

I think we need to discuss this PR more also because @smarterclayton says he had already fixed To to support ImageStreamTag and DockerImage.